### PR TITLE
Fix for Detach Faces undo issue

### PR DESCRIPTION
**Pull request by modrimkus**
**DO NOT FORGET TO INCLUDE A CHANGELOG ENTRY**

### Purpose of this PR

Fixes a bug where undoing after a "Detach faces" operation with the "Separate GameObject" option on would not restore the detached faces - leaving the mesh with missing faces. Behaviour before:

BeforeFix

Behaviour after:

AfterFix

The issue was that the call to `Undo.RegisterCreatedObjectUndo(copy.gameObject, "Detach Selection");` was issued too early without giving the chance for the prior call to  `UndoUtility.RecordSelection("Detach Face(s)");` to register original mesh's face changes.

Tested with small meshes and meshes with >512 vert count as `UndoUtility.RecordSelection` takes different code paths depending on vertex count. In each case, fix worked correctly.

Added automated test.

### Links

**Fogbugz:** https://fogbugz.unity3d.com/f/cases/1407518/

### Comments to Reviewers

[List known issues, planned work, provide any extra context for your code.]

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.probuilder?pullRequest=417) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.probuilder?pullRequest=417) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.probuilder?pullRequest=417) it.</sub>

**Review by JoelFortin — APPROVED** [review]
LGTM

**Review by karl- — APPROVED** [review]

File: Tests/Editor/Undo/TestUndo.cs [hunk 1]
```diff
@@ -36,4 +38,40 @@ public static void RegisterComplete()
...
+        UnityEditor.Undo.PerformUndo();
+
+        // this is usually caught by UndoUtility
+        cube.InvalidateCaches();
+
+        cube.ToMesh();
+        cube.Refresh();
+
+        // After undo, previously edited mesh should match the duplicate
+        TestUtility.AssertAreEqual(duplicate.mesh, cube.mesh);
```
Comment (lopezt-unity) [inline Tests/Editor/Undo/TestUndo.cs:72]: Not sure about that but is it worthy to add a test after detach and before undo to assert the meshes are different?

**Review by lopezt-unity — APPROVED** [review]
LGTM

**Pull request merged**